### PR TITLE
[MRG] MAINT: remove move_to_pos

### LINKED ISSUE #202: scaling of cell positions
I [discovered](https://github.com/jonescompneurolab/hnn-core/pull/200#discussion_r518937845) in #200 that the cell positions are scaled in a somewhat odd way:

https://github.com/jonescompneurolab/hnn-core/blob/bb559b7879c2bf9823d42717438a52b22378e75c/hnn_core/cell.py#L154

This seems like a bug to me but maybe we are missing something. Two questions:

- why is `self.pos[2]` passed instead of `self.pos[1]` for y coordinate
- why is `self.pos[2]` not scaled by 100?

It's not super-clear where this factor 100 comes from. Since the dipole moment = current * distance, this will affect the final result. cc @stephanie-r-jones @rythorpe

### PR DESCRIPTION
**Pull request by jasmainak**
closes #202

### CODE REVIEW

**Comment by jasmainak** [comment]
There is a slight mismatch with original "ground truth" dipole if we do this but I would argue this is more correct. Any thoughts @rythorpe @cjayb ?

**Comment by cjayb** [comment]
It would be super useful to learn something about the reason that was implemented in the first place! It seems utterly bizarre to not treat x and y coords symmetrically. My concern with just removing it is that this seems like yet another tweak that has a specific and thus "intended" purpose, even though it will most likely be wrong for most other simulations with other goals. perhaps something to do with CSD analysis?

**Comment by samnemo** [comment]
You're right that those pt3dchange calls were only written to move the neurons vertically for visualization purposes & shouldn't influence the dipole calculations or circuit connectivity.
________________________________

**Comment by rythorpe** [comment]
I'm fairly certain @jasmainak is right about the discrepancy being caused by rounding error. To demonstrate this, I've printed `self.soma.L` before and after calling `h.pt3dchange()` in the default example in master. As `network_builder` iterates over all cells and moves them into their final positions, I get the following output snippet. Note that [`L2Pyr`](https://github.com/jonescompneurolab/hnn-core/blob/eb173119b1832be6b170c25e738b978e74f05d73/hnn_core/pyramidal.py#L350) cells should have somas that are 22.1 um long and [`L5Pyr`](https://github.com/jonescompneurolab/hnn-core/blob/eb173119b1832be6b170c25e738b978e74f05d73/hnn_core/pyramidal.py#L480) and baskets cells should have somas that are 39.0 um long.

Before/after:
```
22.1 22.0999755859375
22.1 22.0999755859375
22.1 22.0999755859375
22.1 22.0999755859375
22.1 22.0999755859375
39.0 39.0
39.0 39.0
39.0 39.0
39.0 39.0
39.0 39.0
```

Now, if we only move the root compartment (i.e., the soma) of each neuron instead of all sections individually, [NEURON should move the whole tree](https://www.neuron.yale.edu/neuron/static/py_doc/modelspec/programmatic/topology/geometry.html#d-specification-of-geometry) without introducing the rounding error caused by converting floats (except for any rounding error introduced into the soma itself).
```python
    def move_to_pos(self):
        """Move cell to position."""
        x0 = self.soma.x3d(0)
        y0 = self.soma.y3d(0)
        z0 = self.soma.z3d(0)
        dx = self.pos[0] * 100 - x0
        dy = self.pos[2] - y0
        dz = self.pos[1] * 100 - z0
        
        root_sec = self.soma
        for i in range(root_sec.n3d()):
            h.pt3dchange(i, root_sec.x3d(i) + dx, root_sec.y3d(i) + dy,
                         root_sec.z3d(i) + dz, root_sec.diam3d(i),
                         sec=root_sec)
```
As expected, this throws a similar error as this PR, albeit with fewer mismatched elements. I think the difference here is caused by the fact that we still introduced a small amount of rounding error into the somas themselves while avoiding it in all other segments.
```
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 13 / 6801 (0.191%)
E       Max absolute difference: 0.0001
E       Max relative difference: 0.00011343
E        x: array([-4.0000e-04, -4.0000e-04, -4.0000e-04, ...,  9.0972e+00,
E               9.0651e+00,  9.0332e+00])
E        y: array([-4.0000e-04, -4.0000e-04, -4.0000e-04, ...,  9.0972e+00,
E               9.0651e+00,  9.0332e+00])

tests/test_parallel_backends.py:167: AssertionError
```

**Comment by jasmainak** [comment]
@rythorpe @cjayb I actually found a neat way to demo it's a numerical precision issue. Check the last commit 77d4b3116bb436c255159cf9d523945c5c71f79a

What do you think?

**Comment by rythorpe** [comment]
> @rythorpe @cjayb I actually found a neat way to demo it's a numerical precision issue. Check the last commit [77d4b31](https://github.com/jonescompneurolab/hnn-core/commit/77d4b3116bb436c255159cf9d523945c5c71f79a). Tests pass on this
> 
> What do you think?

Sure enough, this makes sense to me. My vote is to move forward with removing the `move_to_pos()` method and update our ground-truth dipole accordingly.

**Comment by cjayb** [comment]
Isn't the correct fix to remove `move_to_pos` entirely, though?

So the conclusion is that `move_to_pos` has lead to slightly wrong currents due to rounding errors? That is, that even though the function was intended to only influence visualisation, it actually also changed the output dipole moment (due to rounding errors)?

This will affect the LFP calculation too.

**Comment by jasmainak** [comment]
yep, I'll remove `move_to_pos` entirely and update the "ground truth". Just wanted to demo it first and be sure that is indeed the issue

**Comment by cjayb** [comment]
Great! This means I'll take a look at removing the `legacy`-flag in `Network`! Now that we'll diverge from legacy GUI output, we'll fly free ;)

**Comment by jasmainak** [comment]
so it looks like Neuron 8.0 also produces some slight differences in results. Not sure it would be straightforward or worth it to track this since none of us are Neuron developers and/or have development version installed. I propose that we accept this slight difference and update the ground truth. What do you guys think?

Question is -- do we update the "ground truth" and require NEURON 8.0 from next release of hnn-core? If we still want to support older NEURON version, we should have two different "ground truths" to compare against. Another possibility is that we could have a "tolerance" parameter, but it's possible that some small bugs pass unnoticed since a tiny change in one missing parameter does not always have a big impact on the simulation output.

All these issues point to the problems of comparing the outputs rather than testing the internals using unit tests. Should we plan to have a "test-a-thon" (hackathon where we write tests) and try to phase out or relax the "ground truth" criteria after that?

**Comment by jasmainak** [comment]
BTW, I just remembered that we used to actually have the dev version of Neuron on Travis in the past. [These are the steps](https://github.com/jonescompneurolab/hnn-core/blob/85698e51a135bc283a42719e3c5f9f990ab87297/.travis.yml#L36-L43) you have to follow. And then git bisect the heck out of it. Anybody wants to give it a shot?

**Comment by rythorpe** [comment]
I think we should resolve this before implementing the updated calcium version. Would it be worth attempting to position each cell to its final destination upon construction (i.e., rather than creating the cell and then moving it), or is the goal to completely get rid of actually positioning the Neuron cell object?

**Comment by cjayb** [comment]
Yes, "real" cell positions are needed for LFP: they are used to calculate the distances to the electrode tips.

**Comment by jasmainak** [comment]
could we make cell spacing an argument to `Network` class so folks can tune it what they think is realistic for LFP simulation? We can throw an error if there are LFP electrodes and the spacing is too large to be realistic.

Instead of changing it through `move_to_pos` we do it in Python itself [here](https://github.com/jonescompneurolab/hnn-core/blob/2f1da11e8f6fdd17f469ce2f5f1376526741d84d/hnn_core/network.py#L23)

**Comment by cjayb** [comment]
How about a `Cell.set_rotation`-method [like LFPy](https://lfpy.readthedocs.io/en/latest/_modules/LFPy/cell.html#Cell.set_rotation)? We'd then do a rotation by `-np.pi` to get them into the XZ-plane.

On the other hand, why not just decide that in `hnn_core`, the pyramidal cell apical axis defines the z-axis? We'd need to change `_secs_LxPyr` for [L5](https://github.com/jonescompneurolab/hnn-core/blob/2f1da11e8f6fdd17f469ce2f5f1376526741d84d/hnn_core/params_default.py#L429) and L2 pyramidals, that's it. Am I missing something there?

My point: if there's no good reason for our cells to lie flat in the XY-plane, why not just re-align them and be done with it?

LFP calculation is affected by the distances, but also geometry: our cells are effectively 2D, as there are only 2 basal dendrites, and the 'oblique' apical stump is in-plane with them.

**Comment by rythorpe** [comment]
> could we make cell spacing an argument to `Network` class so folks can tune it what they think is realistic for LFP simulation? We can throw an error if there are LFP electrodes and the spacing is too large to be realistic.
>
> Instead of changing it through `move_to_pos` we do it in Python itself [here](https://github.com/jonescompneurolab/hnn-core/blob/2f1da11e8f6fdd17f469ce2f5f1376526741d84d/hnn_core/network.py#L23)

I'm working on a different branch at the moment that refactors how cells, their positions, and their gids are assigned. Let's save implementing a cell spacing argument at the network-level for that PR and just resolve the `move_to_pos` dilemma.

What do you think of just changing `Cell.move_to_pos()` to modify [`Cell.p_secs` in the constructor](https://github.com/jonescompneurolab/hnn-core/blob/2f1da11e8f6fdd17f469ce2f5f1376526741d84d/hnn_core/cell.py#L172) so that there is a 1-to-1 comparison between the section points specified in `p_secs` (as accessed by the user in `Network.cell_types` and what is actually implemented in the model?

**Comment by rythorpe** [comment]
+1 for redefining `_secs_LxPyr` so that the apical dendrite extends along the z-axis!

**Comment by jasmainak** [comment]
> On the other hand, why not just decide that in hnn_core, the pyramidal cell apical axis defines the z-axis? We'd need to change _secs_LxPyr for L5 and L2 pyramidals, that's it. Am I missing something there?

If we load a different neuron at some point in the future, we'd have to make sure that the conversion from Neuron y axis to our z-axis is done ...

**Comment by jasmainak** [comment]
> What do you think of just changing Cell.move_to_pos() to modify Cell.p_secs in the constructor so that there is a 1-to-1 comparison between the section points specified in p_secs (as accessed by the user in Network.cell_types) and what is actually implemented in the model?

+1 for this

**Comment by jasmainak** [comment]
@rythorpe do you have any further thoughts here?

**Comment by rythorpe** [comment]
You mean regarding how to handle neurons that are not initially specified with apical dendrites directed along the z-axis?

Re: the more general issue of instantiating cell position and hid in a cleaner manner, I haven’t had a chance to think and/or work on this the last few days as I’ve been traveling, but I’ll give this and a related PR more attention this Thursday.

**Comment by jasmainak** [comment]
okay sounds good! I guess I'm wondering what we decided regarding the y/z coordinate business since the LFP PR depends on it. My initial thought is that we should assume the hard-coded coordinates will be following Neuron convention and then apply the necessary transformations to make it compatible in the `hnn-core` world.

**Comment by rythorpe** [comment]
> okay sounds good! I guess I'm wondering what we decided regarding the y/z coordinate business since the LFP PR depends on it. My initial thought is that we should assume the hard-coded coordinates will be following Neuron convention and then apply the necessary transformations to make it compatible in the `hnn-core` world.

I'm torn here because one of the few black-and-white constraints HNN assumes of its current (or any future) underlying network is the direction and orientation of the apical dendrites of pyramidal cells. Specifically, they should extend outward from the soma normal to the surface of the neocortex, which we ubiquitously visualize at the local level as in the z-direction.

**Comment by cjayb** [comment]
> If we load a different neuron at some point in the future, we'd have to make sure that the conversion from Neuron y axis to our z-axis is done ...

I'm not sure there really is a "Neuron convention"? In any case, a morphology-reader should not assume anything, but be told:

```python
morph_L5 = read_cell_morphology(filename_L5, cell_type='pyramidal', apical_orientation=(0, 1, 0))
net = Network(..., morphology={'L5Pyr': morph_L5, 'L2Pyr': ...})
```

It'll be a while before this becomes relevant, but the above would be pretty transparent, no?

**Comment by jasmainak** [comment]
I suppose Neuron convention is whatever is specified here: https://www.neuron.yale.edu/neuron/static/py_doc/modelspec/programmatic/topology/geometry.html#pt3dadd. If we decide not to use it in the future, we can fly free!

**Comment by jasmainak** [comment]
@rythorpe in terms of plan of action, I think step 1 is to get this PR into mergeable step so we can remove `pt3dadd`. This would involve updating / adding data to the `data` branch and making the CIs pass. For step 2, we can make another PR to change the coordinates.

**Comment by rythorpe** [comment]
> I think step 1 is to get this PR into mergeable step so we can remove `pt3dadd`. This would involve updating / adding data to the `data` branch and making the CIs pass. For step 2, we can make another PR to change the coordinates.

@jasmainak you mean remove `h.pt3dchange` right? We'll need to still initialize the points to create Neuron morphology somehow, we just will initialize them to their final position upon creation so that we're not moving them around in Neuron-space.

Am I clear to rebase and push?

**Comment by jasmainak** [comment]
yep you can rebase and push. I am not touching this PR.

remember that the positions in space don't affect the dipole calculation. They only affect the LFP calculation ...

**Comment by rythorpe** [comment]
Sure, but the plan is still to maintain consistent positioning between what is specified in `Cell.pos` and the corresponding Neuron sections right?

**Comment by jasmainak** [comment]
right, I suppose you'd just have to modify this line:

https://github.com/jonescompneurolab/hnn-core/blob/5c1d2a2c81002d749c6aabd0ebcb199d6095bdfb/hnn_core/cell.py#L259

to:

```py
h.pt3dadd(self.pos[0] + pt[0], 
          self.pos[1] + pt[2],  # Neuron z-coordinate is matplotlib y-coordinate
          self.pos[2] + pt[1], 1, sec=sec)
```

or some such thing

**Comment by rythorpe** [comment]
I was originally thinking of updating `self.p_secs`, but I think your method above will be cleaner especially for visualization.

What's the best way to update the data? PR to the "test_data" branch?

**Comment by jasmainak** [comment]
Yep, PR to "test_data" branch would be the best way!

**Comment by jasmainak** [comment]
can you rebase this PR @rythorpe ?

**Comment by jasmainak** [comment]
anything else to do for the PR to be MRG?

**Comment by rythorpe** [comment]
I rebased last night, have we merged to master since then?

**Comment by jasmainak** [comment]
oh indeed, I restarted the CI to redownload the new test data

**Comment by rythorpe** [comment]
Once CI's turn green, I think we're good to go.

**Comment by rythorpe** [comment]
I'm going to go ahead and merge since tests are passing and this PR is something we've all weighed in on.

**Comment by jasmainak** [comment]
Perfect, thanks @rythorpe for taking it to the finish line!

**Comment by rythorpe** [comment]
You did almost all of the real work here @jasmainak! Stayed tuned for a follow up PR re: cell spacing.

File: hnn_core/cell.py [hunk 1]
```diff
@@ -250,13 +250,22 @@ def _create_sections(self, p_secs, topology):
...
         for sec_name in p_secs:
             sec = h.Section(name=f'{self.name}_{sec_name}')
             self.sections[sec_name] = sec
 
             h.pt3dclear(sec=sec)
             for pt in p_secs[sec_name]['sec_pts']:
-                h.pt3dadd(pt[0], pt[1], pt[2], 1, sec=sec)
+                h.pt3dadd(pt[0] + dx,
+                          pt[1] + dy,
+                          pt[2] + dz, 1, sec=sec)
```
Comment (cjayb) [inline hnn_core/cell.py:268]: The consequences of these lines are very confusing to me (still: I realise nothing has changed in that respect in this PR). Are we reorienting the apical sections to align with the Z-axis here?

In any case, we now have a situation where `net.pos_dict`, `Cell.pos` and the `Section` end points are inconsistent, right? 

@rythorpe are you planning to address this in the follow-up you mention? I'm working on some qualified opinions about the spacing-issue too, based on "reduced" model literature.

Comment (jasmainak) [reply hnn_core/cell.py:268]: yes, we are reorienting the apical sections to align with the Z-axis ... see the lines above.

The `Section` end points are specified relative to the `Cell.pos` or `net.pos_dict` ...

Comment (jasmainak) [reply hnn_core/cell.py:268]: now that I think of it, to reorient along the Z-axis, you'd need to do:

```py
dy = self.pos[1] - self.p_secs['soma']['sec_pts'][0][2]
```

what might be more clear though is to apply a `trans` matrix first:

```py
trans = [[1, 0, 0], [0, 0, 1], [0, 1, 0]]
self.p_secs['soma']['sec_pts'] = np.dot(trans, self.p_secs['soma']['sec_pts'])
```

there is a wealth of functions we can copy from MNE :) There is even a function to align along a target axis:

https://github.com/mne-tools/mne-python/blob/b3f4b05b3933a3b73458205c7e786166f90ad1f7/mne/transforms.py#L298 

that we can use to our advantage

Comment (rythorpe) [reply hnn_core/cell.py:268]: I'm not sure I understand your question @cjayb. Are you asking about the implementation in this PR or what we plan to do in later PRs?

If we change the section endpoints in `params_default.py`, I think we should do that in a separate PR.

**Pull request merged**